cmd/cloud-endpoints-controller/main.go

```diff
@@ -126,84 +130,51 @@ func sync(parent *CloudEndpoint, children *CloudEndpointControllerRequestChildre
-						}
-						if svc.Spec.Type == corev1.ServiceTypeNodePort && len(svc.Spec.Ports) > 0 {
-							nodePort := strconv.Itoa(int(svc.Spec.Ports[0].NodePort))
-							found := false
-							for _, be := range ingBackends {
-								if strings.Contains(be, fmt.Sprintf("k8s-be-%s", nodePort)) {
-									bePatterns[i] = be
-									backend, err := config.clientCompute.BackendServices.Get(config.Project, be).Do()
-									if err == nil {
-										found = true
-										jwtAud := makeJWTAudience(config.ProjectNum, strconv.Itoa(int(backend.Id)))
-										log.Printf("[INFO][%s] Created jwtAud: %s", parent.Name, jwtAud)
-										status.JWTAudiences = append(status.JWTAudiences, jwtAud)
-									}
-								}
-							}
-							if found == false {
-								return status, &desiredChildren, fmt.Errorf("Backend not found or is not ready for service: %s, NodePort: %s", svcName, nodePort)
-							}
-						} else {
-							return status, &desiredChildren, fmt.Errorf("Service %s not type NodePort", svcName)
-						}
-					}
-				}
-			} else {
-				target = parent.Spec.Target
-			}
-			openAPISpec, err = makeDefaultOpenAPISpec(status.Endpoint, target, status.JWTAudiences)
-			if err != nil {
+		if parent.Spec.TargetIngress.Name != "" {
+			target, status.JWTAudiences, err = getTargetIngress(parent)
+			if err != nil { //the ingress isn't ready
+				log.Printf("[INFO][%s] target not ready.", parent.Name)
 				return status, &desiredChildren, err
 			}
-			log.Printf("[INFO][%s] Target: %s", parent.Name, target)
-			status.IngressIP = target
 		} else {
-			// Use provided spec.
-			specYaml, err := yaml.Marshal(&parent.Spec.OpenAPISpec)
+			target = parent.Spec.Target
+		}
+		status.IngressIP = target
+
+		if name, key := parent.Spec.OpenAPISpecConfigMap.Name, parent.Spec.OpenAPISpecConfigMap.Key; name != "" && key != "" {
+			openAPISpecTemplate, err = getConfigMapSpecData(parent.ObjectMeta.Namespace, name, key)
+			if err != nil { //The user tried to supply a configMap spec, but it could not be loaded yet
+				log.Printf("[INFO][%s] Waiting for ConfigMap with Spec named '%s' containing key: '%s'", parent.Name, name, key)
+				return status, &desiredChildren, nil
+			} else {
+				status.ConfigMapHash = toSha1(openAPISpecTemplate)
+			}
+		} else if parent.Spec.OpenAPISpec != nil {
+			openAPISpecTemplateBytes, err := yaml.Marshal(parent.Spec.OpenAPISpec)
 			if err != nil {
-				return status, &desiredChildren, err
+				log.Printf("[ERROR][%s] Supplied spec: %v", parent.Name, err)
+				return status, &desiredChildren, errors.Wrap(err, "Could not load user supplied API from CloudEndpoint Spec")
 			}
+			openAPISpecTemplate = string(openAPISpecTemplateBytes)
+		}
 
-			var spec map[string]interface{}
-			if err = yaml.Unmarshal([]byte(specYaml), &spec); err != nil {
-				return status, &desiredChildren, err
-			}
-			openAPISpec = string(specYaml)
+		if openAPISpecTemplate == "" { // no spec from manifest of configmap
```

Comment:
This should be the else condition to the above conditional block. if you get this far then no spec was given in either the configmap or the parent spec.